src/Microsoft.ML.Parquet/ParquetLoader.cs

```diff
@@ -390,11 +391,32 @@ public Cursor(ParquetLoader parent, Func<int, bool> predicate, IRandom rand)
                     Columns = _loader._columnsLoaded.Select(i => i.Name).ToArray()
                 };
 
-                int numBlocks = (int)Math.Ceiling(((decimal)parent.GetRowCount() / _readerOptions.Count));
-                int[] blockOrder = _rand == null ? Utils.GetIdentityPermutation(numBlocks) : Utils.GetRandomPermutation(rand, numBlocks);
-                _blockEnumerator = blockOrder.GetEnumerator();
+                // The number of blocks is calculated based on the specified rows in a block (defaults to 1M).
+                // Since we want to shuffle the blocks in addition to shuffling the rows in each block, checks
+                // are put in place to ensure we can produce a shuffle order for the blocks.
+                int numBlocks;
+                int[] blockOrder;
+                try
+                {
+                    numBlocks = checked((int)Math.Ceiling(((decimal)parent.GetRowCount() / _readerOptions.Count)));
+                }
+                catch (OverflowException)
+                {
+                    // this exception is thrown when number of blocks exceeds int.MaxValue
+                    throw _loader._host.ExceptParam("ColumnChunkReadSize", "Error due to too many blocks. Try increasing block size.");
+                }
+                try
+                {
+                    blockOrder = _rand == null ? Utils.GetIdentityPermutation(numBlocks) : Utils.GetRandomPermutation(rand, numBlocks);
```

Comment:
>Utils.GetIdentityPermutation(numBlocks) [](start = 49, length = 39)

`Enumerable.Range`. Don't materialize the array in the sequential case...

Reply:
Hi Gleb, this comment was on the sequential case. There's certainly no need to allocate there. Did you mean this comment to go on the above thread?

Let's talk about it here though just so we don't break threads. Right, so, these questions we're having are, I think, a pretty clear sign of something goes well beyond the scope of what this PR is meant to do. I think using a PRNG scheme with as bad of a reputation as LCG or its close variants is not something we ought to just accept reflexively. If you feel strongly about it, perhaps an issue should be filed, so whether or not that level of shuffling is sufficient or produces good practical results can be evaluated. In the meantime, maybe we should entertain the possibility that using the existing mechanism that is used elsewhere without incident may be sufficient?

---
In reply to: [187506292](https://github.com/dotnet/machinelearning/pull/120#discussion_r187506292) [](ancestors = 187506292,187505300)

Reply:
Thank for the clarification Tom. I agree that this discussion, although good to have, is beyond the scope of this PR. The original implementation was done to conform to the pattern of the other shuffling mechanisms of this library. If there is an agreement on the implementation we're happy to follow guidance. At this moment however, I'm seconding the vote for array shuffling until a sufficient MathUtils replacement is created that supports IEnumerator. Please advise.